runtime/convertValues.go

```diff
@@ -1382,6 +1451,90 @@ func (i valueImporter) importDictionaryValue(
 	), nil
 }
 
+func (i valueImporter) importInclusiveRangeValue(
+	v cadence.InclusiveRange,
+	expectedType sema.Type,
+) (
+	*interpreter.CompositeValue,
+	error,
+) {
+	// start, end and step. The order matters.
+	members := make([]interpreter.IntegerValue, 3)
+
+	var memberType sema.Type
+
+	inclusiveRangeType, ok := expectedType.(*sema.InclusiveRangeType)
+	if ok {
+		memberType = inclusiveRangeType.MemberType
+
+		// The inner type must be an integer.
+		if !memberType.Tag().BelongsTo(sema.IntegerTypeTag) {
+			return nil, errors.NewDefaultUserError(
+				"cannot import inclusiverange: member type must be an integer",
+			)
+		}
+	}
+
+	inter := i.inter
+	locationRange := i.locationRange
+
+	// import members.
+	for index, value := range []cadence.Value{v.Start, v.End, v.Step} {
+		importedValue, err := i.importValue(value, memberType)
+		if err != nil {
+			return nil, err
+		}
+		importedIntegerValue, ok := importedValue.(interpreter.IntegerValue)
+		if !ok {
+			return nil, err
+		}
+
+		members[index] = importedIntegerValue
+	}
+
+	// start, end and step. The order matters.
+	memberTypes := make([]sema.Type, 3)
+	memberStaticTypes := make([]interpreter.StaticType, 3)
+
+	for i, member := range members {
+		memberStaticTypes[i] = member.StaticType(inter)
+		memberType, err := inter.ConvertStaticToSemaType(memberStaticTypes[i])
+		if err != nil {
+			return nil, err
+		}
+		memberTypes[i] = memberType
+	}
```

Comment:
Maybe move this snippet also inside the `if inclusiveRangeType == nil {` condition (and merge it with the one below), so it's clear that this code is only needed when the expected range type is unknown

Reply:
I've refactored the function a bit. Can you please take a look again and comment if this is still relevant.

The key point is that I need to ensure each individual field (start, end and step) has the same static type. I've added a comment in the code as to why that is necessary.